### PR TITLE
Initial Readystatus from API-Server

### LINKED ISSUE #78733: pod with readinessProbe will be not ready when kubelet restart
**What happened:**
pod with readinessProbe will be not ready a little while when kubelet restart, and then the endpoint will changed.
**What you expected to happen:**
When kubelet restart, the ready status of the pod should be true rather than false.
**How to reproduce it (as minimally and precisely as possible):**
1 start a pod with readinessProbe
2 kubectl get po <podName> -o yaml --watch      watch the pod status
3 service kubelet restart       restart the kubelet
4 you will see below (for example)
```
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: <time>
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: <time>
    message: 'containers with unready status: [kubedns]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: <time>
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: <containerID>
    image:<image>
    imageID: <imageID>
    name: <kubedns>
    ready: false
    restartCount: 0
    state:
      running:
        startedAt: <time>
```

**Anything else we need to know?:**
No
**Environment:**
Kubernetes v1.10.2

### PR DESCRIPTION
**Pull request by dimakow**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
When kubelet is restarted and kubelet is running syncPod for first time, it is not considering readiness probe state at all. Call stack on first run is:

https://github.com/kubernetes/kubernetes/blob/355bc3df228d9529a15e53d4bb59757e191c0225/pkg/kubelet/kubelet.go#L1556

https://github.com/kubernetes/kubernetes/blob/355bc3df228d9529a15e53d4bb59757e191c0225/pkg/kubelet/kubelet_pods.go#L1425

https://github.com/kubernetes/kubernetes/blob/355bc3df228d9529a15e53d4bb59757e191c0225/pkg/kubelet/prober/prober_manager.go#L240-L259

And Line 259 is the problem here, as it just returns the initialValue to the upper calling layer which will get persisted to API-Server and then cause some minor service-outages because Pods once where notReady although this state never reflected actual container status.

Changes to overcome this problem:

1. StatusManager in starting phase will populate all podStatuses with status from API-Server
2. ProberManager will recognize that readinessManager has not reported a status yet. In this case we will consult statusManager and give back latest ready state from API-Server. This state from API-Server should be returned only for short amount of time as readinessManager should eventually report something

I also took the chance import changes from #79555 as it prevents that a wrong initial ready state is set to readinessManager. The aim of this PR is clearly that when readiness probes report something, they do this with correct state and not some default, non true value as it can cause service disruptions.

Happy to dicuss if this PR is the right approach :)

#### Which issue(s) this PR fixes:
Fixes #78733

#### Does this PR introduce a user-facing change?
NONE

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
Can you add a node e2e that demonstrates the original problem and verifies its solved with this solution?

File: pkg/kubelet/status/status_manager.go
Comment (matthyx) [inline pkg/kubelet/status/status_manager.go:209]: I am afraid this will put strain on the api-manager...
Maybe you could restrict this to local pods with:
```
metav1.ListOptions{FieldSelector: "spec.nodeName=" + nodeName}
```

File: pkg/kubelet/prober/worker.go
Comment (matthyx) [inline pkg/kubelet/prober/worker.go:211]: Can you just explain the logic, and maybe put a comment or extract the if to clearly understand when we should (not) set the initial value?

File: pkg/kubelet/prober/prober_manager.go
Comment (matthyx) [inline pkg/kubelet/prober/prober_manager.go:247]: Isn't it already initialized at `false`?

**Review by matthyx — CHANGES_REQUESTED** [review]
In general I like the idea, but I'm not sure it's realistic... especially on the strain applied to the api-server.
Also if you could add an e2e test validating exactly to this feature, thanks!

File: pkg/kubelet/status/status_manager.go
Comment (ehashman) [reply pkg/kubelet/status/status_manager.go:209]: We should already have a client for the API server somewhere in the status manager/kubelet, I agree that we shouldn't be adding a new one here, especially not to fetch all pods (not even limited to the current node!). This will add unnecessary load.

**Review by ehashman — CHANGES_REQUESTED** [review]
/hold

Per my comment on https://github.com/kubernetes/kubernetes/issues/100277#issuecomment-910517011 I don't actually think this is a bug and I am not sure this change makes sense.

File: pkg/kubelet/status/status_manager.go
Comment (dimakow) [reply pkg/kubelet/status/status_manager.go:209]: I considered the load on the API-Server. It should be ok since statusmanager receives the kubeclient from kubelet and kubelet initializes kubeclient with informers in conjunction with a node filter

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L410-L417

Maybe with sharedInformers we just preload the pods which are requested from kubelet later on. Haven't checked this but at least the nodefilter should be already there.

File: pkg/kubelet/prober/prober_manager.go
Comment (dimakow) [reply pkg/kubelet/prober/prober_manager.go:247]: You are right, this could be improved and is maybe leftover from development.

File: pkg/kubelet/prober/worker.go
Comment (dimakow) [reply pkg/kubelet/prober/worker.go:211]: Problem is that these lines of code are always executed nomatter what. Reason is that w.containerID is not filled upon creation of worker but only here by copying from c.ContainerID

So there needed to be some condition upon which we actually set the initialValue. This PR aims to improve readiness check so we exclude other probeTypes. It might be worth improving this worker.go in general by creating seperate worker_types based on their proberType with their own coding.

The other conditions cover truly new containers which are started for first time, they can be set to initialValue as it is no harm for them.

In general, upon first run worker.go will call resultmanager twice today. Once for filling w.ContainerID and another time for the actual readiness check. I wanted to get rid of the first call because it is stupid to call resultManager with a value which was not dervied from actual testing.

If desired, i can add a comment.

File: pkg/kubelet/prober/prober_manager.go
Comment (dimakow) [reply pkg/kubelet/prober/prober_manager.go:254]: **What if it hasn't?**

Imagine a one replica persistent database behind a service having an endpoint, which will get deleted and added again only when you update the kubelet on the node -  a cascade of load on the API-Server and downtime for consumers of that service.

Nodes, which host one replica persistent databases, should run as long as possible hosting the workload since you cannot easily move around the workload without producing downtime. To make them run as long as possible, you will exchange kubelet from time to time to be on a recent version. Just by default considering all pods not-ready on startup does not sound desirable to me.

Maybe it is worth discussing if this can be enabled optionally?

**What if status has changed?**
If status has changed kubelet will eventually report the actual-ready state. Actually, we only report back the state from API-Server as long as no worker has done it's probe job.

**Comment by dims** [comment]
this PR has been on hold for about 2 weeks. Can you please see how to unblock this by reviewing comments etc? or please close this out (Trying to declutter our open list of PRs)

**Comment by dims** [comment]
This PR is older than 4 weeks and needs as rebase.

Please rebase the PR against latest master and reopen if still needed.

/close

File: pkg/kubelet/prober/prober_manager.go
Comment (vlerenc) [reply pkg/kubelet/prober/prober_manager.go:254]: @ehashman Unfortunately, you have not replied anymore. We are struck by the same.

**Comment by vlerenc** [comment]
@ehashman @bg-chun Why did you let this PR die?

It is a real problem to assume something is dead when one doesn't know, especially for brownfield/singleton workloads or when multiple Kubelets are touched at the same time.

I understand that on startup of the pod, the state of the probe must first be successful and that makes a lot of sense, but if the observer (Kubelet) gets restarted, it should never assume this has changed. Whatever the state was previously is very probably also the state now. By the same logic, you could set it to failed one second after you have set it to successful, because you haven't checked in that second and won't check for many more seconds. When the observer isn't checking, the default is not failed until the timeout lapses and KCM takes action <- that's your safeguard. That the observer breaks workload on purpose, just because it restarted, is as wrong as it gets.

Let me ask another question: Would you build an airplane or rocket engine like that? Switch off one of the engines, if you don't know its status, because some controller or whatever reboots/reinitialises? And if by chance you are too heavy or another engine has a real issue, you let the airplane/rocket crash?

Are you willing to reconsider or is that it now and you leave the Kubelet with this wrong implementation?

**Pull request closed without merging**
